src/PluginManager/Microsoft.Performance.Toolkit.PluginManager.Core/Discovery/DiscoveryManager.cs

```diff
@@ -0,0 +1,175 @@
+        private readonly IDiscovererProvidersLoader loader;
+        private readonly Dictionary<IPluginSource, IList<IPluginDiscoverer>> sourceToDiscoverers;
+
+        public DiscoveryManager(
+            IEnumerable<IPluginDiscovererProvider> discovererProviders,
+            IDiscovererProvidersLoader discovererSourceLoader)
+        {
+            this.pluginSources = new HashSet<IPluginSource>();
+            this.loader = discovererSourceLoader;
+            this.discovererProviders = new HashSet<IPluginDiscovererProvider>(discovererProviders);
+            this.sourceToDiscoverers = new Dictionary<IPluginSource, IList<IPluginDiscoverer>>();
+        }
+
+        public IEnumerable<IPluginSource> PluginSources
+        {
+            get
+            {
+                return this.pluginSources;
+            }
+        }
+
+        public IEnumerable<IPluginDiscovererProvider> DiscovererProviders
+        {
+            get
+            {
+                return this.discovererProviders;
+            }
+        }
+
+        public void ClearPluginSources()
+        {
+            this.pluginSources.Clear();
+            this.sourceToDiscoverers.Clear();
+        }
+
+        public void AddPluginSource<TSource>(TSource source) where TSource : class, IPluginSource
+        {
+            Guard.NotNull(source, nameof(source));
+
+            this.pluginSources.Add(source);
+            this.sourceToDiscoverers.Add(source, new List<IPluginDiscoverer>());
+
+            foreach (IPluginDiscovererProvider<TSource> provider in this.discovererProviders.OfType<IPluginDiscovererProvider<TSource>>())
+            {
+                if (provider.IsSourceSupported(source))
+                {
+                    this.sourceToDiscoverers[source].Add(provider.CreateDiscoverer(source));
+                }
+            }
+        }
+
+        public void LoadAdditionalDiscovererSource(string directory)
+        {
+            Guard.NotNull(directory, nameof(directory));
+
+            if (!this.loader.TryLoad(directory, out IEnumerable<IPluginDiscovererProvider> loadedProviders))
+            {
+                return;
+            }
+
+            IEnumerable<IPluginDiscovererProvider> newProviders = loadedProviders.Except(this.discovererProviders);
+            this.discovererProviders.UnionWith(newProviders);
+
+            foreach (IPluginDiscovererProvider source in newProviders)
+            {
+                Type sourceType = source.PluginSourceType;
+                Type discovererSourceType = source.GetType();
+
+                foreach (KeyValuePair<IPluginSource, IList<IPluginDiscoverer>> kvp in this.sourceToDiscoverers)
+                {
+                    IPluginSource pluginSource = kvp.Key;
+                    if (sourceType !=  pluginSource.GetType())
+                    {
+                        continue;
+                    }
+
+                    if ((bool)discovererSourceType.GetMethod("IsSourceSupported")
```

Comment:
might be good to use "nameof" for these, helps make sure they are tracked if we change them